modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformaceStatisticsProcessor.java

```diff
@@ -0,0 +1,232 @@
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.IgniteFeatures;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.processors.GridProcessorAdapter;
+import org.apache.ignite.internal.util.distributed.DistributedProcess;
+import org.apache.ignite.internal.util.future.GridFinishedFuture;
+import org.apache.ignite.internal.util.future.GridFutureAdapter;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.lang.IgniteFuture;
+import org.apache.ignite.lang.IgniteFutureCancelledException;
+import org.apache.ignite.spi.discovery.DiscoveryDataBag;
+
+import static org.apache.ignite.internal.GridComponent.DiscoveryDataExchangeType.PERFORMANCE_STAT_PROC;
+import static org.apache.ignite.internal.IgniteFeatures.allNodesSupports;
+import static org.apache.ignite.internal.util.distributed.DistributedProcess.DistributedProcessType.PERFORMANCE_STATISTICS;
+
+/**
+ * Performance statistics processor.
+ * <p>
+ * Manages collecting statistics.
+ */
+public class PerformaceStatisticsProcessor extends GridProcessorAdapter {
+    /** Process to start/stop statistics. */
+    private final DistributedProcess<Boolean, Boolean> proc;
+
+    /** Performance statistics writer. */
+    private final FilePerformanceStatistics writer;
+
+    /** Synchronization mutex for request futures. */
+    private final Object mux = new Object();
+
+    /** Enable/disable statistics request futures. */
+    private final ConcurrentMap<UUID, GridFutureAdapter<Void>> reqFuts = new ConcurrentHashMap<>();
+
+    /** Disconnected flag. */
+    private volatile boolean disconnected;
+
+    /** Stopped flag. */
+    private volatile boolean stopped;
+
+    /** @param ctx Kernal context. */
+    public PerformaceStatisticsProcessor(GridKernalContext ctx) {
+        super(ctx);
+
+        writer = new FilePerformanceStatistics(ctx);
+
+        proc = new DistributedProcess<>(ctx, PERFORMANCE_STATISTICS, start -> {
+            if (start) {
+                return ctx.closure().callLocalSafe(() -> {
+                    if (start)
+                        writer.start();
+
+                    return true;
+                });
+            }
+
+            return writer.stop().chain(f -> true);
+        }, (uuid, res, err) -> {
+            if (!F.isEmpty(err) && statisticsEnabled())
+                writer.stop();
+
+            synchronized (mux) {
+                GridFutureAdapter<Void> fut = reqFuts.get(uuid);
+
+                if (fut != null) {
+                    if (!F.isEmpty(err))
+                        fut.onDone(new IgniteException("Unable to process request [err=" + err + ']'));
+                    else
+                        fut.onDone();
+                }
+            }
+        });
+    }
+
+    /** @return Performance statistics writer. */
+    public IgnitePerformanceStatistics writer() {
```

Comment:
Let's encapsulate writer and introduce methods of this interface directly in the processor:
We may want to filter out some event later or have several writers, etc.
These changes should go directly in the processor and don't bother writer or core code.



```
                    ctx.performanceStatistics().task(
                        ses.getId(),
                        ses.getTaskName(),
                        ses.getStartTime(),
                        U.currentTimeMillis() - ses.getStartTime(),
                        worker.affPartId());
```

Reply:
Writer encapsulated.